Dashboard Page

JimFung opened the pull request:
Completes #451 

A couple things to pay attention to: 
1) my css is probably trash, please let me know how I can make it better
2) following up on point 1, I have no idea how to make the list of user actions wrap nicely. open to ideas. (see attached photo)
3) right now the links in the list of actions link directly to the article itself instead of the editor. Not sure if that was the intention or not. 
4) I kept everything inside dashboardpage.js but i don't mind breaking it out into separate components. 
5) Don't think there is a "New Page" action yet, so it doesn't link to anything right now.

<img width="1680" alt="screen shot 2017-02-12 at 5 34 37 pm" src="https://cloud.githubusercontent.com/assets/2354516/22868435/cac3cd3e-f14a-11e6-851c-af48e6dbd52b.png">

psiemens committed:
Expose actions in API

JimFung committed:
finished structure of jsx

JimFung committed:
changed payload for /api/actions endpoint

JimFung committed:
wip, adding recent articles endpoint

JimFung committed:
wip i need to work on cs416 now

JimFung committed:
wip okay now im going to work on 416

JimFung committed:
list of user history done

JimFung committed:
recent articles done

JimFung committed:
adding some styles

JimFung committed:
finished styling everything, added icons to articles list

blx commented on dispatch/apps/core/actions.py:
```diff
@@ -40,16 +40,18 @@ def list_actions(count=25):
+                meta['author'] = action.person.full_name
+                meta['headline'] = article.headline
+                meta['article_url'] = article.get_absolute_url()
+                meta['count'] = 1 if count == 1 else count
```
isn't this equivalent to
```py
meta['count'] = count

JimFung replied:
why yes, yes it is. 😅

blx commented on dispatch/static/manager/src/styles/components/dashboard.scss:
```diff
@@ -0,0 +1,64 @@
+
+.dashboard_container ul li {
+
+}
```
empty style?

JimFung replied:
forgot to delete this.

blx commented on dispatch/static/manager/src/js/pages/DashboardPage.js:
```diff
@@ -1,14 +1,85 @@
+        {elem.meta.count == 1 ? ` ${elem.meta.action} ` : ` made ${elem.meta.count} ${elem.meta.action} to ` }
+        <a href={elem.meta.article_url}>{elem.meta.headline}</a>
+        <span>{` ${moment(elem.timestamp).from(moment())}`}</span>
+      </li>))
```
re point # 2 -- one approach would be to float the icon left, put the text in an inline-block or block element, and slap a margin-left on it. it should wrap at the same indent:

![screen shot 2017-02-12 at 9 51 39 pm](https://cloud.githubusercontent.com/assets/887504/22872378/9f545eac-f16d-11e6-8298-ddf02b46a6d0.png)

```html
    <style>
        .a { margin-left: 2rem; background: #ddd; max-width: 25rem; margin-top: 1rem; }
        i { float: left; }
    </style>

    <div>
        <i class="fa fa-server"></i>
        <div class="a">It was a bright cold day in April, and the clocks were striking thirteen.</div>
    </div>
    <div>
        <i class="fa fa-feed"></i>
        <div class="a">Winston Smith, his chin nuzzled into his breast in an effort to escape the vile wind, slipped quickly through the glass doors of Victory Mansions, though not quickly enough to prevent a swirl of gritty dust from entering along with him.</div>
    </div>
```
(this is divs, not `li`s, but same principle)

JimFung replied:
Thanks, I'll give this a shot!

JimFung replied:
I tried this but it still didn't wrap nicely 😕 maybe i didn't do it right. I'll try it again tonight, if not i might just stick these inside a table cause that'll work... :crossed_fingers:

JimFung replied:
![image](https://cloud.githubusercontent.com/assets/2354516/22940476/dd1137e0-f296-11e6-9301-eda0fc60d029.png)
Nailed it...kinda, that one fifth from the bottom still doesn't look great but it wraps nicely.

psiemens replied:
I played around with this a little bit and got it working nicely with a flexbox. If you don't mind, I can push a fix.

JimFung replied:
👍 yes please do! i wanna see how you did it.

blx commented on dispatch/static/manager/package.json:
```diff
@@ -54,5 +55,8 @@
     "url": "https://github.com/ubyssey/dispatch/issues"
+  },
+  "dependencies": {
+    "moment": "^2.17.1"
```
is moment the only non dev dependency? why is the min version here different from on line 30 above?

JimFung replied:
hmm not sure how it got into devDependencies since i ran npm with the --save tag. I didn't realize our dependencies were all devDependencies, is there any reason why that is? I believe the convention is to save anything required for our app to function as a dependency so when we install it in a production env (with the --production flag for npm) it'll ignore the devDependencies, which are generally linters, test runners, etc. 

I will remove moment from dependencies and leave the one in the devDependency object, since that is the convention this project is using.

blx replied:
yeah i'd have the same reasoning ✌️ maybe ping @psiemens on this

psiemens replied:
If I recall correctly, there used to be a time when saving deps as `devDependencies` was the norm for Webpack projects (since all of them are bundled anyways), but I think that's changed. I'm cool with moving everything over to `dependencies`.

JimFung committed:
the easy changes

JimFung committed:
realized we actually have a colour palette so i switched the colours

JimFung committed:
also realized we have reset font weights so i switched to those

JimFung committed:
renamed the scss file cause it was causing troubles for the sass compiler

JimFung committed:
fixed alignment issues with the text for the user actions

psiemens commented on dispatch/apps/core/actions.py:
```diff
@@ -40,16 +40,18 @@ def list_actions(count=25):
+                meta['headline'] = article.headline
+                meta['article_url'] = article.get_absolute_url()
+                meta['count'] = count
+                meta['action'] = SINGULAR[action.action] if count == 1 else PLURAL[action.action]
```
I'd prefer to write it like this:

```python
meta = {
    'author': action.person.full_name,
    'headline': article.headline,
    ...etc..
}
```
Just more readable IMO.

JimFung replied:
👍 was hoping for more review on my python cause i don't know the syntax at all. Changing this now.

psiemens commented on dispatch/static/manager/src/js/actions/DashboardActions.js:
```diff
@@ -0,0 +1,24 @@
+
+export function getRecentArticles(token) {
+  return {
+    type: types.FETCH_RECENT,
```
Can we make this a bit more verbose?

```js
types.FETCH_RECENT_ARTICLES
```

psiemens commented on dispatch/static/manager/src/js/api/dispatch.js:
```diff
@@ -155,6 +155,14 @@ var DispatchAPI = {
       return postRequest('tags', null, {name: name}, token)
     }
+  },
+  dashboard: {
```
We're sort of breaking out of REST here, but I'm not totally opposed to that...

I'm just wondering if it then makes sense to combine both endpoints, or at least put them under the same root:

`GET api/dashboard/actions`
`GET api/dashboard/recent`

Thoughts?

JimFung replied:
Honestly didn't give it much thought when I built it, figured it'd be easy to move around once the functionality was there. 

/{page}/{data} is probably the most common pattern and i'm not opposed to it. 

Out of habit, I make my endpoints cause this is how i designed my APIs at work cause its a good way to dig through large nested datasets. 
/{resource}/{id/some_other_dimension}

I think for your suggestion is probably better here so i'll change it to that. 👍

JimFung replied:
Done!

psiemens commented on dispatch/static/manager/src/js/pages/DashboardPage.js:
```diff
@@ -1,14 +1,85 @@
   render() {
+
+    //returns the user history if the data is loaded, else returns an empty list element
+    let historyList = this.props.actions.isLoaded
```
This can be a `const`

psiemens commented on dispatch/static/manager/src/js/pages/DashboardPage.js:
```diff
@@ -1,14 +1,85 @@
+    : <li></li>
+
+    //Returns a list of the user's recent articles if the data is loaded, else returns an empty list element
+    let recentArticlesList = this.props.recent.isLoaded
```
Change to `const`

psiemens commented on dispatch/static/manager/src/js/pages/DashboardPage.js:
```diff
@@ -1,14 +1,85 @@
+    //Returns a list of the user's recent articles if the data is loaded, else returns an empty list element
+    let recentArticlesList = this.props.recent.isLoaded
+    ? this.props.recent.data.map((elem,ind) => {
+      return (
```
Change this to implicit return like above

psiemens commented on dispatch/static/manager/src/js/pages/DashboardPage.js:
```diff
@@ -1,14 +1,85 @@
+import { Link } from 'react-router'
+import { connect } from 'react-redux'
+import * as dashboardActions from '../actions/DashboardActions'
+import moment from 'moment'
```
This is nit-picky, but I like to put all external imports before project imports. Can you move this up one line?

psiemens commented on dispatch/static/manager/src/js/reducers/DashboardReducer.js:
```diff
@@ -0,0 +1,61 @@
+  }
+}
+
+let actionsReducer = (state = initialState.actions, action) => {
```
Change to `const`

psiemens commented on dispatch/static/manager/src/js/reducers/DashboardReducer.js:
```diff
@@ -0,0 +1,61 @@
+  }
+}
+
+let recentReducer = (state = initialState.recent, action) => {
```
Change to `const`

psiemens commented on dispatch/static/manager/src/styles/components/_dashboard.scss:
```diff
@@ -0,0 +1,61 @@
+.dashboard_container {
```
Check out the SASS style guide here: https://github.com/ubyssey/ubyssey-dispatch-theme/wiki/SASS-Style-Guide

Sorry, should have let you know before. Shouldn't be hard to change these styles though

psiemens commented on dispatch/static/manager/src/js/pages/DashboardPage.js:
```diff
@@ -1,14 +1,85 @@
+            <h2>Quick Actions</h2>
+            <div className="dashboard_quick_actions">
+              <ul>
+                <li><button className="pt-button"><Link to={`/articles/new`}>New Article</Link></button></li>
```
There's actually a component for this!

```js
import { LinkButton } from '../components/inputs'

...

<LinkButton to='articles/new'>New Article</LinkButton>
```

psiemens commented on dispatch/static/manager/src/js/pages/DashboardPage.js:
```diff
@@ -1,14 +1,85 @@
+        <i className={`fa fa-${elem.icon}`} aria-hidden="true"></i>
+        {elem.meta.author}
+        {elem.meta.count == 1 ? ` ${elem.meta.action} ` : ` made ${elem.meta.count} ${elem.meta.action} to ` }
+        <a href={elem.meta.article_url}>{elem.meta.headline}</a>
```
Can we change this to link to the article edit page in Dispatch instead of the live version?

You should be able to do something like this:

```js
<LinkButton to={`articles/${elem.meta.article_id}`}>{elem.meta.headline}</LinkButton>
```

You'll just have to add the article ID to the API response first :metal:

JimFung replied:
yep, i'll make the api change as part of the change to make the endpoints /dashboard/{resource}

psiemens reviewed (changes requested):
Everything looks really good for the most part! The functionality is nearly perfect.

JimFung committed:
changed how fields in the meta obj is defined

JimFung committed:
updated package.json

JimFung committed:
changed FETCH_RECENT to FETCH_RECENT_ARTICLES

JimFung committed:
changed order of imports

JimFung committed:
reusing components and keeping things constant

JimFung committed:
changed routes for action and recent to be behind /dashboard

JimFung committed:
links in article actions now link to the editor

JimFung committed:
removed debug code

JimFung committed:
updated styles to follow style guide

psiemens committed:
Small JS refactor, tweak styles

- Move activity and articles lists to separate methods
- Use flex box instead of floats
- Small style tweaks

psiemens commented:
Just made a few updates to the CSS and JSX code. Merging!
